internal/app/machined/pkg/runtime/drainer.go

```diff
@@ -0,0 +1,94 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package runtime
+
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// NewDrainer creates new drainer.
+func NewDrainer() *Drainer {
+	return &Drainer{}
+}
+
+// Drainer is used in controllers to ensure graceful shutdown.
+type Drainer struct {
+	subscriptionsMu sync.Mutex
+	subscriptions   []*DrainSubscription
+}
+
+// Drain initializes drain sequence waits for it to succeed until the context is canceled.
+func (d *Drainer) Drain(ctx context.Context) error {
+	var eg errgroup.Group
+
+	d.subscriptionsMu.Lock()
+	defer d.subscriptionsMu.Unlock()
+
+	for _, s := range d.subscriptions {
+		s := s
+
+		eg.Go(func() error {
```

Comment:
suggestion: we can make s.events length 1 and simply send to each subscriber event without goroutine